internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -397,22 +407,59 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 		// We intentionally discard/ignore this after the last Function runs.
 		fctx = rsp.GetContext()
 
+		for _, c := range rsp.GetConditions() {
+			var status corev1.ConditionStatus
+			switch c.Status {
+			case v1beta1.Status_STATUS_CONDITION_TRUE:
+				status = corev1.ConditionTrue
+			case v1beta1.Status_STATUS_CONDITION_FALSE:
+				status = corev1.ConditionFalse
+			case v1beta1.Status_STATUS_CONDITION_UNKNOWN, v1beta1.Status_STATUS_CONDITION_UNSPECIFIED:
+				status = corev1.ConditionUnknown
+			}
+
+			conditions = append(conditions, TargetedCondition{
+				Condition: xpv1.Condition{
+					Type:               xpv1.ConditionType(c.Type),
+					Status:             status,
+					LastTransitionTime: metav1.Now(),
+					Reason:             xpv1.ConditionReason(c.Reason),
+					Message:            ptr.ToString(c.Message),
+				},
+				Target: convertTarget(c.GetTarget()),
+			})
+		}
+
 		// Results of fatal severity stop the Composition process. Other results
 		// are accumulated to be emitted as events by the Reconciler.
 		for _, rs := range rsp.GetResults() {
+			reason := reasonCompose
+			if rs.Reason != nil {
+				reason = event.Reason(rs.GetReason())
+			}
+
+			target := convertTarget(rs.GetTarget())
+
+			var e event.Event
 			switch rs.GetSeverity() {
 			case v1beta1.Severity_SEVERITY_FATAL:
-				return CompositionResult{}, errors.Errorf(errFmtFatalResult, fn.Step, rs.GetMessage())
+				// TODO(dalton, reviewers): Safe to add events here? This is a behavior
+				// change, but allows function authors to emit events prior to failure.
+				return CompositionResult{Events: events, Conditions: conditions}, errors.Errorf(errFmtFatalResult, fn.Step, rs.GetMessage())
```

Comment:
Good points, I agree that returning a partial result feels wrong.

> We find some way other than returning an error (e.g. a new field in CompositionResult{} to represent a fatal result.

This seems like a fair change IMO. For example the composite reconciler could check for `err != nil || result.FatalResult != nil`. It would then have logic to only handle the results and conditions in the case of `err ==nil && FatalResult != nil`. This is effectively what is happening currently because all other errors do not return `events` or `conditions`.

I don't feel strongly either way, LMK if you end up wanting to change this.